File: syspop/r/base_pop.R
#' Create Base Population
#'
#' This function generates a synthetic base population by repeating rows in the
#' input data frame based on the values specified in the "value" column.
#' It filters the data frame based on the specified area if a filter is provided.
#' The output includes an index column representing the row number.
#'
#' @param structure_data A data frame containing the population structure data.
#'                       It must include a column named "area" and a column named "value".
#' @param output_area_filter An optional vector of area IDs to filter the rows of
#'                           the structure_data. If NULL, no filtering is applied.
#'
#' @return A data frame containing the generated base population with repeated rows.
#'         The "value" column is removed, and a new column "index" is added that
#'         contains the row numbers.
#'
#' @examples
#' # Example of using the base_pop_wrapper function
#' data <- data.frame(area = c(241300, 241300, 241301),
#'                    value = c(12, 9, 15))
#' population <- base_pop_wrapper(data, output_area_filter = c(241300))
#' print(population)
#'
base_pop_wrapper <- function(
    structure_data, 
    output_area_filter = NULL
) {

  # Start the timer
  start_time <- now(tz = "UTC")
  
  # Filter based on output_area_filter if provided
  if (!is.null(output_area_filter)) {
    structure_data <- structure_data[structure_data$area %in% output_area_filter, ]
  }

  # Repeat rows based on the "value" column and drop the "value" column
  if (!is.null(structure_data)) {
    
    population <- structure_data %>%
      as.data.frame() %>%  # Ensure we are working with a data frame
      .[rep(1:nrow(.), times = as.integer(.$value)), ] %>%
      select(-value)  # Remove the value column
    
    rownames(population) <- NULL

    population <- population %>%
      mutate(index = row_number())
  }
  
  # End the timer
  end_time <- now(tz = "UTC")
  
  # Calculate the processing time in minutes
  total_mins <- round(as.numeric(difftime(end_time, start_time, units = "mins")), 2)
  
  # Log the processing time
  print(sprintf("Processing time (base population): %.2f minutes", total_mins))
  
  return(population)
}

File: syspop/python/base_pop.py
from datetime import datetime
from logging import getLogger

from numpy.random import choice
from pandas import DataFrame

logger = getLogger()

def base_pop_wrapper(
    structure_data: DataFrame,
    output_area_filter: list or None,
) -> DataFrame:
    """Create base population

    Args:
        gender_data (DataFrame): Gender data for each age
        ethnicity_data (DataFrame): Ethnicity data for each age
        output_area_filter (list or None): With area ID to be used

    Returns:
        DataFrame: Produced base population

    The output is sth like:
                area  age  gender ethnicity
        0     236800    0  female  European
        1     236800    0    male  European
        2     236800    0    male  European
        3     236800    0  female  European
        4     236800    0    male  European
        ...      ...  ...     ...       ...
    """

    start_time = datetime.utcnow()
    if output_area_filter is not None:
        structure_data = structure_data[structure_data["area"].isin(output_area_filter)]

    if structure_data is not None:
        df_repeated = structure_data.loc[structure_data.index.repeat(structure_data["value"].astype(int))]
        population = df_repeated.reset_index(drop=True).drop(columns=["value"])

    end_time = datetime.utcnow()
    total_mins = (end_time - start_time).total_seconds() / 60.0

    logger.info(f"Processing time (base population): {total_mins}")

    return population
